### DISCUSSION TITLE
Is everyone supposed to be able to upgrade to Next JS 13 right now?

### DISCUSSION THREAD

**Discussion by 0xabrar** [vercel/next.js#50152] [Help] [answered]
### Summary

Looking over the summary it seems like the biggest things are in beta and there's some breaking changes. Is everyone supposed to be able to move over? Or do we kind of have to wait if we're using packages like Emotion/Tailwind/Chakra-UI. I tried upgrading and running the codemods but it didn't really do anything.

I'm getting an error with the following:

`yarn build
uncaughtException /home/dev/Documents/mylinks/node_modules/next/dist/compiled/undici/index.js:1
(()=>{var __webpack_modules__={1900:(A,e,t)=>{"use strict";const{parseContentType:s}=t(4318);function getInstance(A){const e=A.headers;const t=s(e["content-type"]);if(!t)throw new Error("Malformed content type");for(const s of o){const o=s.detect(t);if(!o)continue;const r={limits:A.limits,headers:e,conType:t,highWaterMark:undefined,fileHwm:undefined,defCharset:undefined,defParamCharset:undefined,preservePath:false};if(A.highWaterMark)r.highWaterMark=A.highWaterMark;if(A.fileHwm)r.fileHwm=A.fileHwm;r.defCharset=A.defCharset;r.defParamCharset=A.defParamCharset;r.preservePath=A.preservePath;return new s(r)}throw new Error(`Unsupported content type: ${e["content-type"]}`)}const o=[t(6104),t(8506)].filter((function(A){return typeof A.detect==="function"}));A.exports=A=>{if(typeof A!=="object"||A===null)A={};if(typeof A.headers!=="object"||A.headers===null||typeof A.headers["content-type"]!=="string"){throw new Error("Missing Content-Type")}return getInstance(A)}},6104:(A,e,t)=>{"use strict";const{Readable:s,Writable:o}=t(2781);const r=t(6542);const{basename:i,convertToUTF8:n,getDecoder:g,parseContentType:C,parseDisposition:E}=t(4318);const Q=Buffer.from("\r\n");const I=Buffer.from("\r");const B=Buffer.from("-");function noop(){}const a=2e3;const c=16*1024;const h=0;const l=1;const u=2;class HeaderParser{constructor(A){this.header=Object.create(null);this.pairCount=0;this.byteCount=0;this.state=h;this.name="";this.value="";this.crlf=0;this.cb=A}reset(){this.header=Object.create(null);this.pairCount=0;this.byteCount=0;this.state=h;this.name="";this.value="";this.crlf=0}push(A,e,t){let s=e;while(e<t){switch(this.state){case h:{let o=false;for(;e<t;++e){if(this.byteCount===c)return-1;++this.byteCount;const t=A[e];if(f[t]!==1){if(t!==58)return-1;this.name+=A.latin1Slice(s,e);if(this.name.length===0)return-1;++e;o=true;this.state=l;break}}if(!o){this.name+=A.latin1Slice(s,e);break}}case l:{let o=false;for(;e<t;++e){if(this.byteCount===c)return-1;++this.byteCount;const t=A[e];if(t!==32&&t!==9){s=e;o=true;this.state=u;break}}if(!o)break}case u:switch(this.crlf){case 0:for(;e<t;++e){if(this.byteCount===c)return-1;++this.byteCount;const t=A[e];if(y[t]!==1){if(t!==13)return-1;++this.crlf;break}}this.value+=A.latin1Slice(s,e++);break;case 1:if(this.byteCount===c)return-1;++this.byteCount;if(A[e++]!==10)return-1;++this.crlf;break;case 2:{if(this.byteCount===c)return-1;++this.byteCount;const t=A[e];if(t===32||t===9){s=e;this.crlf=0}else{if(++this.pairCount<a){this.name=this.name.toLowerCase();if(this.header[this.name]===undefined)this.header[this.name]=[this.value];else this.header[this.name].push(this.value)}if(t===13){++this.crlf;++e}else{s=e;this.crlf=0;this.state=h;this.name="";this.value=""}}break}case 3:{if(this.byteCount===c)return-1;++this.byteCount;if(A[e++]!==10)return-1;const t=t`

Which unfortunately doesn't say much at all. Definitely interested in knowing if someone else had this same thing come up.

### Additional information

I was able to upgrade to React 18 so there is at least that.

**Comment by icyJoseph** [comment]
You should be able to upgrade to Next 13, and continue using the pages directory, though you need to fix your usage of `next/link` and `next/image` as well as some middleware tweaks.

To move to Next 13, and use the app directory, you need to first learn how server and client components work, how routing and how data fetching is supposed to work.

- https://nextjs.org/blog/next-13-4#what-does-app-router-stability-mean
- https://nextjs.org/blog/next-13-4#is-the-pages-router-going-away

What did you try to do to upgrade?

**Comment by 0xabrar** [comment]
OK maybe I go through to try and manually refactor link and image, because I ran the codemod on a bunch of different folders and it never seemed to update anything (which was curious). I haven't moved over to the app directory yet.

My thinking was it was more likely something with emotion. I'm still using Babel but it looks like SWC now supports it?

https://nextjs.org/docs/architecture/nextjs-compiler#emotion
https://www.npmjs.com/package/@swc/plugin-emotion

**Comment by 0xabrar** [comment]
Well I upgraded all of the relevant stuff to use SWC instead of babel for compilation (without moving away from using emotion). That works, but still end up the same error. Deleted node_modules and .next folders as well and that didn't seem to do anything.

**Comment by icyJoseph** [comment]
I wonder if you have also upgraded Node.js version? Since Node.js 15, uncaught exceptions terminate the process, rather than just printing a warning, IIRC.

Also undici is a fetch engine, which was used as base for the native Node fetch - odd that it claims to crash there. Perhaps some fetch fails and it is not properly handled, which crashes the process? Have you upgraded your libraries too?

**Comment by pawelmidur** [comment, marked as answer]
@0xabrar install node `>=16` or `>18`

**Reply by 0xabrar** [reply]
Thanks that did it for me :)

I was using nodejs 14, which I thought was still supported (docs only mentioned  EOL for version 12). But updating to use 18 got rid of that error I was having. Still need to go through and fix some bugs and verify that everything works after upgrading to node 18, but those ones are actually legible.

**Reply by 0xabrar** [reply]
Cool everything looks to be still working on the site https://mylinks.ai/ after using the codemods for images + links. Only thing I'd need to do later is update away from legacy images.
